backend/app.ts

```diff
@@ -80,8 +87,67 @@ passport.use(new GoogleStrategy({
         }
     }));
 
+passport.use(new FacebookStrategy({
+    clientID: facebookClientID,
+    clientSecret: facebookAppSecret,
+    callbackURL: "http://localhost:8000/auth/facebook/callback",
+    profileFields: ['id', 'displayName', 'email'],
+    enableProof: true
+},
+    async function verify(accessToken: any, refreshToken: any, profile: any, cb: any) {
+        try {
+            const user = await db.user.findFirst({
+                where: {
+                    facebookID: profile.id
+                }
+            })
+
+            if (!user) {
+                const newUser = await db.user.create({
+                    data: {
+                        firstName: profile._json.name,
+                        lastName: profile._json.name,
+                        facebookID: profile._json.id,
+                        email: profile._json.email
+                    }
+                })
+                cb(null, newUser)
+            } else {
+                cb(null, user)
+            }
+        } catch (error) {
+            cb(error, null)
+        }
+    }
+));
+
+passport.use(new LocalStrategy(async (email: string, password: string, done: Function) => {
+    try {
+        const user = await db.user.findUnique({
+            where: {
+                email
+            },
+        });
+
+        if (!user || !user.salt) {
+            return done(null, false, { message: 'Incorrect email or password.' });
+        }
+
+        const hashedPassword = crypto.pbkdf2Sync(password, user.salt, 310000, 32, 'sha256').toString('hex');
+        if (user.password !== hashedPassword) {
+            return done(null, false, { message: 'Incorrect email or password.' });
+        }
+
+        return done(null, user);
+
+    } catch (error) {
+        done(error);
+    }
```

Comment:
Since you're FLYING and implementing a shit ton of OAUTH. Challenge for you! We're repeating a lot of Code between Google OAUTH, FB OAUTH and local strategy. How could we DRY the code up?

Reply:
One method that I can think of is to find the implementation similarities between those passport auth strategies and then abstract the user finding & creation logic to some helper functions, and reuse those functions inside our passport strategies.

Reply:
For Google and FB OAUTH, I find that they very similar implementations and I can try to refactor them. As for the local strategy, I'm only seeing 1 implementation to authenticate user login with email and hashedpassword, unless we need to reuse this local strategy logic in the future. Can you give me pointers?